source/core_mqtt_serializer.c

```diff
@@ -1785,896 +3414,1985 @@ MQTTStatus_t MQTT_GetConnectPacketSize( const MQTTConnectInfo_t * pConnectInfo,
+        }
+        return status;
     }
-    else
+
+    MQTTStatus_t MQTTPropAdd_ConnMaxPacketSize(MqttPropBuilder_t* pPropertyBuilder, uint32_t maxPacketSize)
     {
-        /* Empty else MISRA 15.7 */
+        MQTTStatus_t status = MQTTSuccess;
+        uint8_t* pIndex;
+        if ((pPropertyBuilder == NULL))
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder=%p.", (void*)pPropertyBuilder));
+            status = MQTTBadParameter;
+        }
+        else if (pPropertyBuilder->pBuffer == NULL)
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder->pBuffer=%p.", (void*)pPropertyBuilder->pBuffer));
+            status = MQTTBadParameter;
+        }
+        else if ((maxPacketSize == 0) || (UINT32_CHECK_BIT(pPropertyBuilder->fieldSet, MQTT_MAX_PACKET_SIZE_POS)))
+        {
+            LogError(("Max packet size already set"));
+            status = MQTTBadParameter;
+        }
+        else
+        {
+            pIndex = pPropertyBuilder->pBuffer + pPropertyBuilder->currentIndex;
+            *pIndex = MQTT_MAX_PACKET_SIZE_ID;
+            pIndex++;
+            pIndex[0] = UINT32_BYTE3(maxPacketSize);
+            pIndex[1] = UINT32_BYTE2(maxPacketSize);
+            pIndex[2] = UINT32_BYTE1(maxPacketSize);
+            pIndex[3] = UINT32_BYTE0(maxPacketSize);
+            pIndex = &pIndex[4];
+            pPropertyBuilder->fieldSet = UINT32_SET_BIT(pPropertyBuilder->fieldSet, MQTT_MAX_PACKET_SIZE_POS);
+            pPropertyBuilder->currentIndex += 5;
+        }
+        return status;
     }
 
-    if( status == MQTTSuccess )
+    MQTTStatus_t MQTTPropAdd_ConnTopicAliasMax(MqttPropBuilder_t* pPropertyBuilder, uint16_t topicAliasMax)
     {
-        if( pFixedBuffer->size < MQTT_DISCONNECT_PACKET_SIZE )
+        uint8_t* pIndex;
+        MQTTStatus_t status = MQTTSuccess; 
+        if ((pPropertyBuilder == NULL))
         {
-            LogError( ( "Buffer size of %lu is not sufficient to hold "
-                        "serialized DISCONNECT packet of size of %lu.",
-                        ( unsigned long ) pFixedBuffer->size,
-                        MQTT_DISCONNECT_PACKET_SIZE ) );
-            status = MQTTNoMemory;
+            LogError(("Arguments cannot be NULL : pPropertyBuilder=%p.", (void*)pPropertyBuilder));
+            status = MQTTBadParameter;
+        }
+        else if (pPropertyBuilder->pBuffer == NULL)
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder->pBuffer=%p.", (void*)pPropertyBuilder->pBuffer));
+            status = MQTTBadParameter;
+        }
+        else if (UINT32_CHECK_BIT(pPropertyBuilder->fieldSet, MQTT_TOPIC_ALIAS_MAX_POS))
+        {
+            LogError(("Topic Alias Maximum already set. "));
+            status = MQTTBadParameter;
+        }
+        else
+        {
+            pIndex = pPropertyBuilder->pBuffer + pPropertyBuilder->currentIndex;
+            *pIndex = MQTT_TOPIC_ALIAS_MAX_ID;
```

Comment:
We need checks to make sure that we are not going out of bound of the `pPropertyBuilder->pBuffer` by verifying `pPropertyBuilder->currentIndex` and the required encoding length with `pPropertyBuilder->bufferLength`.
This can be added as a separate preliminary check for all similar APIs with a separate error code.